monix-eval/shared/src/main/scala/monix/eval/internal/TaskCache.scala

```diff
@@ -0,0 +1,191 @@
+        state.get() match {
+          case p: Promise[A] @unchecked =>
+            // Resetting the state to `null` will trigger the
+            // execution again on next `runAsync`
+            if (state.compareAndSet(p, null)) {
+              p.tryComplete(value)
+              ()
+            } else {
+              // Race condition, retry
+              // $COVERAGE-OFF$
+              cacheValue(value)
+              // $COVERAGE-ON$
+            }
+          case _ =>
+            // $COVERAGE-OFF$
+            () // Do nothing, as value is probably null already
+          // $COVERAGE-ON$
+        }
+      }
+    }
+
+    /** Builds a callback that gets used to cache the result. */
+    private def complete(implicit s: Scheduler): Callback[Throwable, A] =
+      new Callback[Throwable, A] {
+        def onSuccess(value: A): Unit =
+          self.cacheValue(Success(value))
+        def onError(ex: Throwable): Unit =
+          self.cacheValue(Failure(ex))
+      }
+
+    /** While the task is pending completion, registers a new listener
+      * that will receive the result once the task is complete.
+      */
+    private def registerListener(
+        p: Promise[A],
+        context: Context,
+        cb: Callback[Throwable, A]
+      )(
+        implicit ec: ExecutionContext
+      ): Unit = {
+
+      p.future.onComplete { r =>
+        // Listener is cancelable: we simply ensure that the result isn't streamed
+        if (!context.connection.isCanceled) {
+          context.frameRef.reset()
+          startFull(Task.fromTry(r), context, cb, null, null, null, 1)
+        }
+      }
+    }
+
+    /**
+      * Starts execution, eventually caching the value on completion.
+      */
+    @tailrec private def start(context: Context, cb: Callback[Throwable, A]): Unit = {
+      implicit val sc: Scheduler = context.scheduler
+      self.state.get() match {
+        case null =>
+          val update = Promise[A]()
+
+          if (!self.state.compareAndSet(null, update)) {
+            // $COVERAGE-OFF$
+            start(context, cb) // retry
+            // $COVERAGE-ON$
+          } else {
+            // Registering listener callback for when listener is ready
+            self.registerListener(update, context, cb)
+
+            // Running main task in `uncancelable` model
+            val ctx2 = context
+              .withOptions(context.options.disableAutoCancelableRunLoops)
+              .withConnection(TaskConnection.uncancelable)
+
+            // Start with light async boundary to prevent stack-overflows!
+            Task.unsafeStartTrampolined(self.thunk, ctx2, self.complete)
+          }
+
+        case ref: Promise[A] @unchecked =>
+          self.registerListener(ref, context, cb)
+
+        case (cacheTime: Long, ref: Try[A]) =>
+          // Race condition happened
+          // $COVERAGE-OFF$
+          if (sc.clockRealTime(MILLISECONDS) - cacheTime > duration.toMillis) {
```

Comment:
It's easier and possibly more efficient to store the time at which the value is supposed to expire, instead of the current timestamp. It saves you from doing this math operation on every read. Also, reuse `duration.toMillis` if possible.

Reply:
Good idea, now i can recall this from my time with embedded systems. 
I think the added advantage of calculating the expiration time that it is immune to a counter that rolls over.